/**:
    ros__parameters:
        common:
            lid_topic:  "/velodyne_points"
            imu_topic:  "/imu/data" 

        preprocess:
            lidar_type: 2                # Velodyne LiDAR
            scan_line: 32                # LiDAR scan line number
            blind: 0.5                     # blind area in the beginning of each scan line   
            feature_extract_en: false

        calibration:
            cut_frame: true              # false: do not cut input pointcloud
            cut_frame_num: 3             # must be positive integer
            orig_odom_freq: 10           # original Lidar frequency (for velodyne, it is 10Hz)
            mean_acc_norm: 9.81          # 1: for livox built-in IMU

            imu_sensor_height : 0.73      # Height from ground to IMU sensor (meter)
            verbose : true                # For debug, Rotation matrix related ground, lidar sensor height

            # For Data Accumulation
            data_accum_length: 300.0      # To determine how much data is used
            x_accumulate : 0.01           # Parameter that determines how much the x-axis rotates (0.0 ~ 0.999)
            y_accumulate : 0.01           # Parameter that determines how much the y-axis rotates (0.0 ~ 0.999)
            z_accumulate : 0.99           # Parameter that determines how much the z-axis rotates (0.0 ~ 0.999)

            # initial guess for translation vector (from IMU frame I to LiDAR frame L) - unit : m
            trans_IL_x : 0.0 
            trans_IL_y : 0.0
            trans_IL_z : 0.0
            set_boundary : false
            bound_th : 0.3

            gyro_factor : 10.0
            acc_factor : 1.0
            ground_factor : 5.0

        mapping:
            filter_size_surf: 0.5 
            filter_size_map: 0.5 
            gyr_cov: 0.5 
            acc_cov: 0.5 
            det_range: 200.0              # LiDAR range
            ground_cov : 500.0

        publish:
            path_en:  true                # true: output the path of the robot
            scan_publish_en: true         # false: close all the point cloud output
            dense_publish_en: false       # false: low down the points number in a global-frame point clouds scan.
            scan_bodyframe_pub_en: true   # true: output the point cloud scans in IMU-body-frame

        # For LiDAR Map Saving
        pcd_save:
            pcd_save_en: false
            interval: -1                 # how many LiDAR frames saved in each pcd file; 
                                        # -1 : all frames will be saved in ONE pcd file, may lead to memory crash when having too much frames.

        # For LiDAR odometry Saving
        trajectory_save:
            traj_save_en: true
            traj_file_path: "/root/catkin_ws/src/result/traj.txt" # When you use docker, you should enter docker absolute path.

        # For use patchwork++, more info : https://github.com/url-kaist/patchwork-plusplus 
        patchworkpp:
            sensor_height: 0.85        # For M2DGR dataset : https://github.com/SJTU-ViSYS/M2DGR

            mode: "czm" 
            verbose: false   # To check effect of uprightness/elevation/flatness
            visualize: true  # Ground Likelihood Estimation is visualized
            
            # Ground Plane Fitting parameters
            num_iter: 3             # Number of iterations for ground plane estimation using PCA.
            num_lpr: 20             # Maximum number of points to be selected as lowest points representative.
            num_min_pts: 10         # Minimum number of points to be estimated as ground plane in each patch.
            th_seeds: 0.3           # threshold for lowest point representatives using in initial seeds selection of ground points.
            th_dist: 0.125          # threshold for thickenss of ground.
            th_seeds_v: 0.25        # threshold for lowest point representatives using in initial seeds selection of vertical structural points.
            th_dist_v: 0.1          # threshold for thickenss of vertical structure.
            max_r: 4.0             # max_range of ground estimation area
            min_r: 1.0              # min_range of ground estimation area
            uprightness_thr: 0.707  # threshold of uprightness using in Ground Likelihood Estimation(GLE). Please refer paper for more information about GLE.

            adaptive_seed_selection_margin: -1.2 # The points below the adaptive_seed_selection_margin * sensor_height are filtered
            czm:
                num_zones: 4
                num_sectors_each_zone: [16, 32, 54, 32]
                mum_rings_each_zone: [2, 4, 4, 4]
                elevation_thresholds:  [0.1, 0.2, 0.4, 0.6] # threshold of elevation for each ring using in GLE. Those values are updated adaptively.
                flatness_thresholds:  [0.0, 0.0, 0.0, 0.0]  # threshold of flatness for each ring using in GLE. Those values are updated adaptively.

            enable_RNR : true
            enable_RVPF : true
            enable_TGR : true
            
            RNR_ver_angle_thr : -15.0 # Noise points vertical angle threshold. Downward rays of LiDAR are more likely to generate severe noise points.
            RNR_intensity_thr : 0.2   # Noise points intensity threshold. The reflected points have relatively small intensity than others.